neurom/core/_soma.py

```diff
@@ -88,8 +72,95 @@ def __init__(self, points):
+                (repr(self._points), self.center, self.radius))
+
+
+class SomaCylinders(Soma):
+    '''Soma composed of cylinders (like in SWC)
+
+    points describe the locations of the cylinder start/end points, with
+    their respective radii, much like how neurites are described: ex:
+
+                 /)
+                / o)
+        ______ /  )
+      (|)      ) /
+     ( o )    o )
+      (|)_____ )
+
+      Here we have a 'side-view', with each 'o' representing a point, and the
+      radius is the heigh of a '|' character, and the ')' try and show the
+      curvature of the cylinger
+
+  Note: when, as in the case above, the cylinder center points don't lie
+  in a line, then the overlap between cylinders isn't taken into account for
+  the area calculation
+  '''
+    def __init__(self, points):
+        super(SomaCylinders, self).__init__(points)
+        self.area = sum(morphmath.segment_area((p0, p1))
+                        for p0, p1 in zip(points, points[1:]))
+        self.radius = math.sqrt(self.area / (4. * math.pi))
+
+    @property
+    def center(self):
+        '''Obtain the center from the first stored point'''
+        return self._points[0][COLS.XYZ]
+
+    def __str__(self):
+        return ('SomaCylinders(%s) <center: %s, virtual radius: %s>' %
+                (repr(self._points), self.center, self.radius))
+
+
+class SomaThreePointCylinders(SomaCylinders):
+    ''' NeuroMorpho compatible soma
+
+    Reference:
+        http://neuromorpho.org/SomaFormat.html
+
+        Quote (Tue Feb 28 14:26:56 CET 2017):
+        The first point constitutes the center of the soma, with coordinates (xs, ys,
+        zs) corresponding to the average of all the points in the single contour. An
+        equivalent radius (rs) is computed as the average distance of each point of the
+        single contour from this center. The first point is assigned parent ID -1 as is
+        standard for the root in the SWC format. The second and third soma points, as
+        well as all starting points (roots) of dendritic and axonal arbors have this
+        first point as the parent (parent ID 1). The x and z coordinates as well as the
+        radius of both the second and third soma points are the same as that of the
+        first point. The y coordinates of the second and third points are shifted by -rs
+        and +rs, respectively (Figure 2). The surface area of this soma cylinder equals
+        the surface area of a sphere of radius rs.
+    '''
+    def __init__(self, points):
+        super(SomaThreePointCylinders, self).__init__(points)
+
+        # X    Y     Z   R    P
+        # xs ys      zs rs   -1
+        # xs (ys-rs) zs rs    1
+        # xs (ys+rs) zs rs    1
+
+        # make sure the above invariant holds
+        assert points[0, COLS.R] == points[1, COLS.R] == points[2, COLS.R], \
+            'All radii must be the same'
+        assert points[0, COLS.Y] - points[1, COLS.Y] == points[0, COLS.R], \
```

Comment:
so, in *theory*, the spec suggests they should match exactly, but doing a tolerance check makes sense to me, so I'll add that.